keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -1061,28 +1062,90 @@ Details to consider:
 dereference particular fields.
 
 We can support this in much the same way as cel-policy-template `terms`. These
 can be lazily evaluated while the validation expressions are evaluated
-(cel-policy-template does this). The results can also be memoized to avoid
-repeated evaluations if they are shared across validations.
+(cel-policy-template does this).
+
+A policy can include an additional `variables` section. This is an array
+containing one or more `name` and `expression` pairs, which can be used/re-used by
+the policy's validation expressions. These results are memoized on a
+per-validation basis, so if multiple expressions use the same spec variables,
+the expression that calculates the variable's value will only be ran once.
+
+The variables can be accessed as members of `variables`, which is an object
+that is exposed to CEL expressions (both validation expressions as well as
+other variables).
+
+For example:
 
 ```yaml
   variables:
     - name: metadataList
       expression: "spec.list.map(x, x.metadata)"
     - name: itemMetadataNames
-      expression: "metadataList.map(m, m.name)"
+      expression: "variables.metadataList.map(m, m.name)"
   validations:
-    - expression: "itemMetadataNames.all(name, name.startsWith('xyz-'))"
-    - expression: "itemMetadataNames.exists(name, name == 'required')"
+    - expression: "variables.itemMetadataNames.all(name, name.startsWith('xyz-'))"
+    - expression: "variables.itemMetadataNames.exists(name, name == 'required')"
 ```
 
+Variable names must be valid CEL names. What constitutes a
+valid CEL name can be found at CEL's [language definition](https://github.com/google/cel-spec/blob/master/doc/langdef.md#syntax) under `IDENT`.
+This validity is checked at write time.
+
+For per-policy runtime cost limit purposes, variables count towards the runtime cost limit
+once per policy. The cost of each variable is computed when it is first evaluated in an
+expression, mirroring how the cost limit would be calculated if the variable's
+expression was embedded verbatim.  If the runtime cost limit is exceeded in the
+process, then evaluation halts. No individual variable or expression will be listed as the
+cause in the resulting message. Whether or not the request actually fails depends on the failure policy,
+however. For subsequent uses, inclusion of the variable has zero effect on the runtime
+cost limit. If the variable evaluates to an array or some other iterable, and some expression
+iterates on it, that of course contributes to the cost limit, but simply including the variable does
+not add the underlying expression's cost again.
+
+Variables are also subject to the per-expression runtime cost limit. Exceeding the per-expression
+runtime cost limit is always attributed to the variable, unlike the per-policy limit.
+
+If a variable evaluates to a CEL type at compile-time, then that information will be
+used for any type-checking done on expressions where the variable is used.
+
```

Comment:
Suggest to remove any info regarding with informational type checking. We will add additional info if needed later for informational type checking

Reply:
+1 Let's leave them as dyn for now.